geo/src/algorithm/kmeans/mod.rs

```diff
@@ -0,0 +1,1213 @@
+        // Test with infinite coordinates
+        let points = [
+            point!(x: 0.0, y: 0.0),
+            point!(x: f64::INFINITY, y: 1.0),
+            point!(x: 2.0, y: 2.0),
+        ];
+        let result = points.kmeans(2);
+
+        // Should return InitializationFailed error due to infinity
+        assert!(matches!(
+            result,
+            Err(KMeansError::InitializationFailed(
+                KMeansInitError::InfiniteCoordinate
+            ))
+        ));
+    }
+
+    #[test]
+    fn test_kmeans_reproducibility_with_seed() {
+        // Test that using the same seed produces identical results
+        let points = [
+            point!(x: 0.0, y: 0.0),
+            point!(x: 1.0, y: 0.0),
+            point!(x: 0.0, y: 1.0),
+            point!(x: 10.0, y: 10.0),
+            point!(x: 11.0, y: 10.0),
+            point!(x: 10.0, y: 11.0),
+        ];
+
+        let params = KMeansParams::new(2).seed(42);
+
+        let result1 = points.kmeans_with_params(params.clone()).unwrap();
+        let result2 = points.kmeans_with_params(params).unwrap();
+
+        // Same seed should produce identical results
+        assert_eq!(result1, result2);
+    }
+
+    #[test]
+    fn test_kmeans_reproducibility_with_seed_and_max_radius() {
+        // Test that seed produces reproducible results even when max_radius triggers splitting
+        // Create a cluster that will exceed max_radius
+        let points = [
+            point!(x: 0.0, y: 0.0),
+            point!(x: 100.0, y: 0.0),
+            point!(x: 0.0, y: 100.0),
+            point!(x: 100.0, y: 100.0),
+            point!(x: 50.0, y: 50.0),
+        ];
+
+        let params = KMeansParams::new(1).seed(42).max_radius(40.0); // Will trigger splitting since centroid is at (50, 50) and max point distance is ~70
+
+        let result1 = points.kmeans_with_params(params.clone()).unwrap();
+        let result2 = points.kmeans_with_params(params).unwrap();
+
+        // Same seed should produce identical results even with max_radius splitting
+        assert_eq!(result1, result2);
+    }
+
+    #[test]
+    fn test_kmeans_builder_pattern() {
+        // Test the builder pattern with all options
+        let points = [
+            point!(x: 0.0, y: 0.0),
+            point!(x: 1.0, y: 0.0),
+            point!(x: 0.0, y: 1.0),
+            point!(x: 10.0, y: 10.0),
+            point!(x: 11.0, y: 10.0),
+            point!(x: 10.0, y: 11.0),
+        ];
+
+        // Use builder pattern to configure all parameters
+        let params = KMeansParams::new(2).seed(42).max_iter(100).tolerance(0.001);
+
+        let labels = points.kmeans_with_params(params).unwrap();
+
+        // Should successfully cluster into 2 groups
+        let unique_labels: std::collections::HashSet<_> = labels.iter().copied().collect();
+        assert_eq!(unique_labels.len(), 2);
+    }
+
+    #[test]
+    fn test_kmeans_converges_with_sufficient_iterations() {
```

Comment:
Based on the other tests, we already know that kmeans *can* work. A more interesting test would see it **fail** to converge. I briefly tried such a test, but it didn't fail (presumably convergence happened quickly).


```
        // Use low max_iter which is not enough to converge
        let params = KMeansParams::new(2).seed(42).max_iter(1);
        let result = points.kmeans_with_params(params);

        // Should converge successfully
        assert!(result.is_err());
```

Similarly, a test that actually uses the max_radius parameter to produce interesting results would be nice - the current max_radius test is basically a "it doesn't explode" test.